include/engine/datafacade/bigramblock_datafacade_base.hpp

```diff
@@ -158,325 +158,265 @@ class SharedDataFacade final : public BaseDataFacade
 
         util::ShM<GraphNode, true>::vector node_list(
-            graph_nodes_ptr, data_layout->num_entries[storage::SharedDataLayout::GRAPH_NODE_LIST]);
+            graph_nodes_ptr, data_layout->num_entries[storage::DataLayout::GRAPH_NODE_LIST]);
         util::ShM<GraphEdge, true>::vector edge_list(
-            graph_edges_ptr, data_layout->num_entries[storage::SharedDataLayout::GRAPH_EDGE_LIST]);
+            graph_edges_ptr, data_layout->num_entries[storage::DataLayout::GRAPH_EDGE_LIST]);
         m_query_graph.reset(new QueryGraph(node_list, edge_list));
     }
 
-    void LoadNodeAndEdgeInformation()
+    void InitNodeAndEdgeInformation(storage::DataLayout *data_layout, char *memory_block)
     {
         const auto coordinate_list_ptr = data_layout->GetBlockPtr<util::Coordinate>(
-            shared_memory, storage::SharedDataLayout::COORDINATE_LIST);
-        m_coordinate_list.reset(
+            memory_block, storage::DataLayout::COORDINATE_LIST);
+
+        m_coordinate_list = CoordinateArrayRef(
             coordinate_list_ptr,
-            data_layout->num_entries[storage::SharedDataLayout::COORDINATE_LIST]);
+            coordinate_list_ptr + data_layout->num_entries[storage::DataLayout::COORDINATE_LIST]);
 
         for (unsigned i = 0; i < m_coordinate_list.size(); ++i)
         {
             BOOST_ASSERT(GetCoordinateOfNode(i).IsValid());
         }
 
-        const auto osmnodeid_list_ptr = data_layout->GetBlockPtr<std::uint64_t>(
-            shared_memory, storage::SharedDataLayout::OSM_NODE_ID_LIST);
-        m_osmnodeid_list.reset(
+        const auto osmnodeid_list_ptr = data_layout->GetBlockPtr<OSMNodeID>(
+            memory_block, storage::DataLayout::OSM_NODE_ID_LIST);
+        m_osmnodeid_list = OSMNodeArrayRef(
             osmnodeid_list_ptr,
-            data_layout->num_entries[storage::SharedDataLayout::OSM_NODE_ID_LIST]);
-        // We (ab)use the number of coordinates here because we know we have the same amount of ids
-        m_osmnodeid_list.set_number_of_entries(
-            data_layout->num_entries[storage::SharedDataLayout::COORDINATE_LIST]);
+            osmnodeid_list_ptr + data_layout->num_entries[storage::DataLayout::COORDINATE_LIST]);
 
         const auto travel_mode_list_ptr = data_layout->GetBlockPtr<extractor::TravelMode>(
-            shared_memory, storage::SharedDataLayout::TRAVEL_MODE);
+            memory_block, storage::DataLayout::TRAVEL_MODE);
         util::ShM<extractor::TravelMode, true>::vector travel_mode_list(
-            travel_mode_list_ptr, data_layout->num_entries[storage::SharedDataLayout::TRAVEL_MODE]);
+            travel_mode_list_ptr, data_layout->num_entries[storage::DataLayout::TRAVEL_MODE]);
         m_travel_mode_list = std::move(travel_mode_list);
 
-        const auto lane_data_id_ptr = data_layout->GetBlockPtr<LaneDataID>(
-            shared_memory, storage::SharedDataLayout::LANE_DATA_ID);
+        const auto lane_data_id_ptr =
```

Comment:
all of these following blocks should really be abstracted over, reducing the duplicate behaviour and reducing possible chances for errors.